### PR TITLE
spirv-fuzz: Do not add constants for Block-decorated structs

### LINKED ISSUE #3794: spirv-fuzz: Do not generate constants for decorated struct types
If a struct type is decorated with Block or similar, it's better if we do not generate constants of this type.  Various shader compiler crashes have been detected relating to this, but they are of questionable interest as it's such an edge case.

### PR DESCRIPTION
**Pull request by afd**
Fixes #3794.

### CODE REVIEW

File: source/fuzz/transformation_add_constant_composite.cpp [hunk 1]
```diff
@@ -72,6 +72,24 @@ bool TransformationAddConstantComposite::IsApplicable(
       }
       break;
     case SpvOpTypeStruct:
+      // We do not create constants of structs decorated with Block nor
+      // BufferBlock.  The SPIR-V spec does not explicitly disallow this, but it
+      // seems like a strange thing to do, so we disallow it to avoid triggering
+      // low priorty edge case issues related to it.
+      if (!ir_context->get_decoration_mgr()->WhileEachDecoration(
+              composite_type_instruction->result_id(), SpvDecorationBlock,
+              [](const opt::Instruction & /*unused*/) -> bool {
```
Comment (Vasniktel) [inline source/fuzz/transformation_add_constant_composite.cpp:81]: A minor issue: `const opt::Instruction &` -> `const opt::Instruction&`.

File: source/fuzz/transformation_add_constant_composite.cpp [hunk 2]
```diff
@@ -72,6 +72,24 @@ bool TransformationAddConstantComposite::IsApplicable(
...
+                return false;
+              })) {
+        return false;
+      }
+      if (!ir_context->get_decoration_mgr()->WhileEachDecoration(
+              composite_type_instruction->result_id(), SpvDecorationBufferBlock,
+              [](const opt::Instruction & /*unused*/) -> bool {
```
Comment (Vasniktel) [inline source/fuzz/transformation_add_constant_composite.cpp:88]: A minor issue: `const opt::Instruction &` -> `const opt::Instruction&`.

File: source/fuzz/transformation_add_constant_composite.cpp (hunk 1 above)
Comment (Vasniktel) [reply source/fuzz/transformation_add_constant_composite.cpp:81]: Also, you can adjust comments in the header file for the `IsApplicable` method.

**Review by Vasniktel — COMMENTED** [review]
Looks good. I left some minor suggestions.

File: source/fuzz/transformation_add_constant_composite.cpp (hunk 1 above)
Comment (afd) [reply source/fuzz/transformation_add_constant_composite.cpp:81]: Weirdly clang-format doesn't like what you're proposing (I guess it introduced the space).  If I remove the "/*unused*/" comment then clang-format doesn't want the space, but with the comment it does!

Comment (afd) [reply source/fuzz/transformation_add_constant_composite.cpp:81]: Comment in IsApplicable updated.

File: source/fuzz/transformation_add_constant_composite.cpp (hunk 2 above)
Comment (afd) [reply source/fuzz/transformation_add_constant_composite.cpp:88]: Same as above.

File: source/fuzz/transformation_add_constant_composite.cpp (hunk 1 above)
Comment (Vasniktel) [reply source/fuzz/transformation_add_constant_composite.cpp:81]: In that case, I'm ok with whatever you think is best.

File: source/fuzz/transformation_add_constant_composite.cpp [hunk 3]
```diff
@@ -72,6 +72,24 @@ bool TransformationAddConstantComposite::IsApplicable(
       }
       break;
     case SpvOpTypeStruct:
+      // We do not create constants of structs decorated with Block nor
+      // BufferBlock.  The SPIR-V spec does not explicitly disallow this, but it
+      // seems like a strange thing to do, so we disallow it to avoid triggering
+      // low priorty edge case issues related to it.
+      if (!ir_context->get_decoration_mgr()->WhileEachDecoration(
+              composite_type_instruction->result_id(), SpvDecorationBlock,
```
Comment (Vasniktel) [inline source/fuzz/transformation_add_constant_composite.cpp:80]: You can use the following pattern here:
```cpp
for (auto decoration : {SpvDecorationBlock, SpvDecorationBufferBlock}) {
  ...
}
```

**Review by Vasniktel — APPROVED** [review]

File: source/fuzz/transformation_add_constant_composite.cpp (hunk 3 above)
Comment (afd) [reply source/fuzz/transformation_add_constant_composite.cpp]: Done!  My bet is that the older MSVC compilers will complain about lambda capture though.

**Review by Vasniktel — APPROVED** [review]

**Pull request merged**